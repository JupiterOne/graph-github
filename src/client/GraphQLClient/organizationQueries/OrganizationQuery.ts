import { ExecutableQuery, QueryExecutor } from '../CreateQueryExecutor';
import { OrgQueryResponse, RateLimitStepSummary } from '../types';

const buildQuery = (login: string): ExecutableQuery => {
  const query = `
    query ($login: String!) {
      organization(login: $login) {
        id
        login
        name
        createdAt
        updatedAt
        description
        email
        databaseId
        isVerified
        location
        websiteUrl
        url
      }
      ...rateLimit
    }`;

  return {
    query,
    queryVariables: {
      login,
    },
  };
};

export type OrganizationResults = {
  rateLimit: RateLimitStepSummary;
  organization: OrgQueryResponse;
};

const processResponseData = (responseData): OrganizationResults => {
  if (!responseData) {
    throw new Error('responseData param is required');
  }

  const rateLimit = responseData.rateLimit;
  const organization = responseData.organization;

  return {
    rateLimit: {
      ...rateLimit,
      totalCost: rateLimit?.cost ?? 0,
    },
    organization,
  };
};

/**
 * Fetches organization associated with the provided login.
 * @param login - aka organization
 * @param execute
 */
const fetchOrganization = async (
  login: string,
  execute: QueryExecutor,
): Promise<OrganizationResults> => {
  const executable = buildQuery(login);
  const response = await execute(executable);

  return processResponseData(response);
};

export default { fetchOrganization };
