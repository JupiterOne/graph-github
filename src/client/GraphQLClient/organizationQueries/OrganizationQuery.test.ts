import OrganizationQuery from './OrganizationQuery';

describe('OrganizationQuery', () => {
  test('#fetchOrganization', async () => {
    const response = {
      organization: {
        id: 'O_kgDOBfl5ZQ',
        login: 'j1-ingest',
        name: 'j1-ingest',
        createdAt: '2022-02-22T19:40:36Z',
        updatedAt: '2022-03-23T20:28:45Z',
        description: 'This is a test Org for J1 Ingestion',
        email: 'test@test.com',
        databaseId: 100235621,
        isVerified: false,
        location: 'United States of America',
        websiteUrl: null,
        url: 'https://github.com/j1-ingest',
      },
      rateLimit: {
        cost: 1,
        limit: 5000,
        resetAt: '2022-03-24T18:44:44Z',
        remaining: 4999,
      },
    };
    const executor = jest.fn().mockResolvedValueOnce(response);

    const result = await OrganizationQuery.fetchOrganization(
      'j1-test',
      executor,
    );

    expect(result.rateLimit).toEqual({ ...response.rateLimit, totalCost: 1 });
    expect(result.organization).toEqual(response.organization);
    expect(executor.mock.calls[0][0]).toMatchSnapshot();
  });

  test('#fetchOrganization without rateLimit', async () => {
    const response = {
      organization: {
        id: 'O_kgDOBfl5ZQ',
        login: 'j1-ingest',
        name: 'j1-ingest',
        createdAt: '2022-02-22T19:40:36Z',
        updatedAt: '2022-03-23T20:28:45Z',
        description: 'This is a test Org for J1 Ingestion',
        email: 'test@test.com',
        databaseId: 100235621,
        isVerified: false,
        location: 'United States of America',
        websiteUrl: null,
        url: 'https://github.com/j1-ingest',
      },
    };
    const executor = jest.fn().mockResolvedValueOnce(response);

    const result = await OrganizationQuery.fetchOrganization(
      'j1-test',
      executor,
    );

    expect(result.rateLimit).toEqual({ totalCost: 0 });
    expect(result.organization).toEqual(response.organization);
  });
});
