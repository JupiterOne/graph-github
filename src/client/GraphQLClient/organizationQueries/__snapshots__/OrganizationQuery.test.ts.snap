// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationQuery #fetchOrganization 1`] = `
Object {
  "query": "
    query ($login: String!) {
      organization(login: $login) {
        id
        login
        name
        createdAt
        updatedAt
        description
        email
        databaseId
        isVerified
        location
        websiteUrl
        url
      }
      ...on Query {
    rateLimit {
      limit
      cost
      remaining
      resetAt
    }
  }
    }",
  "queryVariables": Object {
    "login": "j1-test",
  },
}
`;
